tools/common/Rewriter.cs

```diff
@@ -328,17 +364,17 @@ IEnumerable<TypeDefinition> AllTypes (TypeDefinition type)
 			}
 		}
 
-		string ToOutputFileName (string pathToInputFileName)
-		{
-			return Path.Combine (outputDirectory, Path.GetFileName (pathToInputFileName));
-		}
-
 		void MarkForSave (AssemblyDefinition assembly)
 		{
-			var annotations = linkContext.Annotations;
-			var action = annotations.GetAction (assembly);
-			if (action == AssemblyAction.Copy)
-				annotations.SetAction (assembly, AssemblyAction.Save);
+			var annotations = derivedLinkContext.Annotations;
+			if (!annotations.HasAction (assembly)) {
+				annotations.SetAction (assembly, AssemblyAction.Link);
+			} else {
+				var action = annotations.GetAction (assembly);
+				if (action != AssemblyAction.Link) {
+					annotations.SetAction (assembly, AssemblyAction.Link);
```

Comment:
This doesn't look quite right, we shouldn't suddenly link assemblies that we weren't asked to link:

```suggestion
					annotations.SetAction (assembly, AssemblyAction.Save);
```

Reply:
I'll change it, but oddly enough it doesn't make a difference. In the linker pipeline in the b/f save step if the assembly is marked either link or save it gets saved.